test/src/unit-capi-smoke-test.cc

```diff
@@ -0,0 +1,757 @@
+      dimension_ranges.emplace_back(dim_buffer_len);
+    }
+
+    // Create the dimension write buffers.
+    // NOTE: if there is more than 1 dimension, the buffer shall be the previous
+    // buffer multiplied by the current range
+    for (const uint64_t range : dimension_ranges) {
+      uint64_t* write_buffer;
+      if (range == *dimension_ranges.begin()) {
+        write_buffer = (uint64_t*)tdb_malloc(range * sizeof(uint64_t*));
+      } else {
+        write_buffer =
+            (uint64_t*)tdb_malloc(range * (range - 1) * sizeof(uint64_t*));
+      }
+      d_write_buffers.emplace_back(write_buffer);
+    }
+
+    // Fill the write buffers with values.
+    auto dims_iter = test_dims.begin();
+    for (uint64_t* const d_write_buffer : d_write_buffers) {
+      for (uint64_t i = 0; i < dimension_ranges[i]; i++) {
+        d_write_buffer[i] = i;
+      }
+      write_buffer_size = dim_buffer_len * sizeof(uint64_t);
+
+      write_query_buffers.emplace_back(
+          dims_iter->name_,
+          d_write_buffer,
+          &(write_buffer_size),
+          nullptr,
+          nullptr);
+      dims_iter = std::next(dims_iter, 1);
+    }
+
+    // Execute the write query.
+    write(array_name, write_query_buffers, write_order, encryption_type);
+
+    // Free the dimension write buffers vector
+    d_write_buffers.clear();
+    d_write_buffers.shrink_to_fit();
+  } else {
+    // Execute the write query.
+    write(array_name, write_query_buffers, write_order, encryption_type);
+  }
+
+  // Define the read query buffers for "a".
+  vector<test_query_buffer_t> read_query_buffers;
+
+  uint64_t a_read_buffer_size;
+  int32_t* a_read_buffer = nullptr;
+
+  if (test_attr.cell_val_num_ == TILEDB_VAR_NUM) {
+    a_read_buffer_size =
+        total_cells * 2 * tiledb_datatype_size(test_attr.type_);
+    a_read_buffer = (int32_t*)tdb_malloc(a_read_buffer_size);
+    for (uint64_t i = 0; i < total_cells * 2; i++) {
+      a_read_buffer[i] = 0;
+    }
+  } else {
+    a_read_buffer_size = total_cells * tiledb_datatype_size(test_attr.type_);
+    a_read_buffer = (int32_t*)tdb_malloc(a_read_buffer_size);
+    for (uint64_t i = 0; i < total_cells; i++) {
+      a_read_buffer[i] = 0;
+    }
+  }
+
+  uint64_t a_read_buffer_offset_size = total_cells * sizeof(uint64_t);
+  uint64_t* a_read_buffer_offset =
+      (uint64_t*)tdb_malloc(a_read_buffer_offset_size);
+  for (uint64_t i = 0; i < total_cells; i++) {
+    a_read_buffer_offset[i] = 0;
+  }
+
+  // Variable buffer is set only when the attribute is var-sized.
+  if (test_attr.cell_val_num_ == TILEDB_VAR_NUM) {
+    read_query_buffers.emplace_back(
+        test_attr.name_,
+        a_read_buffer_offset,
+        &a_read_buffer_offset_size,
+        a_read_buffer,
+        &a_read_buffer_size);
+  } else {
+    read_query_buffers.emplace_back(
+        test_attr.name_, a_read_buffer, &a_read_buffer_size, nullptr, nullptr);
+  }
+
+  uint64_t subarray_len = 2 * test_dims.size();
```

Comment:
please add an in-code comment that this logic assumes all dimensions are of type UINT64, which is currently true